sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms.splittabledofn;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.range.ByteKey;
+import org.apache.beam.sdk.io.range.ByteKeyRange;
+import org.apache.beam.sdk.transforms.DoFn;
+
+/**
+ * A {@link RestrictionTracker} for claiming {@link ByteKey}s in a {@link ByteKeyRange} in a
+ * monotonically increasing fashion.
+ */
+public class ByteKeyRangeTracker extends RestrictionTracker<ByteKeyRange, ByteKey> {
+  private ByteKeyRange range;
+  @Nullable private ByteKey lastClaimedKey = null;
+  @Nullable private ByteKey lastAttemptedKey = null;
+
+  private ByteKeyRangeTracker(ByteKeyRange range) {
+    this.range = checkNotNull(range);
+  }
+
+  /**
+   * Instantiates a new {@link ByteKeyRangeTracker} with the specified range. The keys in the range
+   * are left padded to be the same length in bytes.
+   */
+  public static ByteKeyRangeTracker of(ByteKeyRange range) {
+    return new ByteKeyRangeTracker(ByteKeyRange.of(range.getStartKey(), range.getEndKey()));
+  }
+
+  @Override
+  public synchronized ByteKeyRange currentRestriction() {
+    return range;
+  }
+
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey, range);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
+    // must always be less than *end*.
+    ByteKey rangeEndKey =
```

Comment:
Agree, but [ByteKeyRange checks that end will always be after the start](https://github.com/apache/beam/blob/3eacf8b8d7301d1bd4db17874a31caf3b9949526/sdks/java/core/src/main/java/org/apache/beam/sdk/io/range/ByteKeyRange.java#L286). I do prefer to remove that restriction from ByteKeyRange or at least loosen it to include equal start and end. Do you agree if I do this? I am not sure this will impact the existing code but we can try.

Reply:
Or we should just allow checkpoint method to return null if a proper residual cannot be provided (since the range has been fully consumed). If we choose that path we should clarify in the interface. @jkff WDYT ?

Reply:
Yeah I'm in favor of loosening the restriction - that seems much less invasive than changing semantics of checkpoint().